### PR TITLE
net/bluetooth: Add iob buffer count and hci receive

### PR DESCRIPTION
**Pull request by amitpurwar**
Signed-off-by: amit purwar <amit.purwar@samsung.com>

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [60bd81c8abba0f42cd9f4051045a43ceb0ef01e2] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [60bd81c8abba0f42cd9f4051045a43ceb0ef01e2] - Code Rule Check OK.

File: os/net/bluetooth/bt_hcicore.h [hunk 1]
```diff
@@ -341,10 +341,10 @@ int bt_hci_cmd_send_sync(uint16_t opcode, FAR struct bt_buf_s *buf, FAR struct b
  * not multi-threading safe
  */
 
-#ifdef CONFIG_DEBUG_WIRELESS_ERROR
```
Comment (dh79pyun) [inline os/net/bluetooth/bt_hcicore.h]: Just remove the line instead of comment out

File: os/include/tinyara/bluetooth/bt_hci.h [hunk 2]
```diff
@@ -219,7 +219,7 @@
 #define BT_HCI_LE_ADV_DISABLE                   0x00
 #define BT_HCI_LE_ADV_ENABLE                    0x01
 #define CONFIG_BT_ID_MAX                        2
-#define HCI_MAX_NAME_LENGTH             248
+#define HCI_MAX_NAME_LENGTH                     65
```
Comment (dh79pyun) [inline os/include/tinyara/bluetooth/bt_hci.h]: Is this change required? When we use '248' length, which problem is occured?

Comment (amitpurwar) [reply os/include/tinyara/bluetooth/bt_hci.h]: * BLUETOOTH_MAX_FRAMELEN
 * Maximum amount of data that can fit in a buffer.
 *
 * The biggest foreseeable buffer size requirement right now comes from
 * the Bluetooth 4.2 SMP MTU which is 65. This then become 65 + 4 (L2CAP
 * header) + 4 (ACL header) + 1 (H4 header) = 74. This also covers the
 * biggest HCI commands and events which are a bit under the 70 byte
 * mark.
 */

#define BLUETOOTH_MAX_FRAMELEN  (BLUETOOTH_MAX_MTU + BLUETOOTH_MAX_HDRLEN)

#define BLUETOOTH_MAX_MTU       70
#define BLUETOOTH_L2CAP_HDRLEN  4       /* Size of L2CAP header */
#define BLUETOOTH_ACL_HDRLEN    4       /* Size of ACL header */
#define BLUETOOTH_H4_HDRLEN     1       /* Size of H4 header */

so bt_buf could be maximum 79 byte.
when I was testing write local name with 248 byte, that time, there was crash happening.
So I found out that we can not have local name with 248 byte, we can have local name with 65 byte and together with header (9 byte), bt_buf would be 74 byte.

**Comment by seinfra** [comment]
Target : [df0c97060bc65edde6c06b0040d9835b0cf1f212] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [df0c97060bc65edde6c06b0040d9835b0cf1f212] - Code Rule Check OK.

File: os/include/tinyara/bluetooth/iob/iob.h [hunk 3]
```diff
@@ -66,6 +66,7 @@
 #endif
 
 /* Some I/O buffers should be allocated */
+#define CONFIG_IOB_NBUFFERS 20
```
Comment (pillip8282) [inline os/include/tinyara/bluetooth/iob/iob.h]: In TizenRT CONFIG_* should be generated in Kconfig. you can't use 'CONFIG_' prefix to macro which defined in source codes. So move it to Kconfig if it need to be changed

File: os/net/bluetooth/bluetooth.c [hunk 4]
```diff
@@ -328,8 +312,11 @@ static int bt_init(void)
 		return ret;
 	}
 
-	bt_finalize_init();
+#ifdef CONFIG_BLUETOOTH_NULL
+	write_local_name("TizenRT");
```
Comment (Taejun-Kwon) [inline os/net/bluetooth/bluetooth.c]: I don't know what is local name and why it needed but bt_null is lower device driver of bluetooth, upper doesn't know which device is attached directly. (of course we can add some process to get capability of lower device through the ioctl)
So you have to find another way.
But you already call bt_set_name in bt_enable, why write it once again??

File: os/net/bluetooth/bluetooth.c [hunk 5]
```diff
@@ -328,8 +312,11 @@ static int bt_init(void)
...
+#endif
 
+	bt_finalize_init();
```
Comment (Taejun-Kwon) [inline os/net/bluetooth/bluetooth.c]: 'finalize' is not good naming, currently bt_finalize_init make bt as a ready state, then bt_set_ready or something will be better.

File: os/net/bluetooth/bluetooth.c (hunk 4 above)
Comment (amitpurwar) [reply os/net/bluetooth/bluetooth.c]: It is written for testing purpose with bt_null driver.

Comment (amitpurwar) [reply os/net/bluetooth/bluetooth.c]: I have removed the write local  name call from bt_init

**Comment by seinfra** [comment]
Target : [a9991a989b7930edfa61ac2e8d1208b2a7417946] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [a9991a989b7930edfa61ac2e8d1208b2a7417946] - Code Rule Check OK.

File: os/include/tinyara/bluetooth/iob/iob.h (hunk 3 above)
Comment (amitpurwar) [reply os/include/tinyara/bluetooth/iob/iob.h]: Moved the  CONFIG_IOB_NBUFFERS to Kconfig

File: os/net/bluetooth/bluetooth.c (hunk 5 above)
Comment (amitpurwar) [reply os/net/bluetooth/bluetooth.c]: This function is from Zephyr's opensource, and we are porting Zephyr's code into TizenRT
https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/host/hci_core.c

File: os/net/bluetooth/bt_netdev.c [hunk 6]
```diff
@@ -69,6 +69,13 @@
 #include "bt_conn.h"
 #include "bt_ioctl.h"
 
+#ifdef CONFIG_BLUETOOTH_NULL
+int bt_dev_register(FAR const struct bt_driver_s *btdev)
```
Comment (Taejun-Kwon) [inline os/net/bluetooth/bt_netdev.c]: Actually register bt dev is not related to CONFIG_BLUETOOTH_NULL.
Below is correct example.
```
#ifdef CONFIG_BLUETOOTH_XXX //regarding net dev
bt_driver *bt_lower;
#ifdef CONFIG_BLUETOOTH_NULL
   bt_driver_s *bt_lower = bt_null_blahblah();
#elif CONFIG_BLUETOOTH_ANOTHER_MODEL
   bt_driver_s *bt_ lwer = bt_another_model_init_blahblah();
#else
   LOG("No lower device available, do not register here\n");
    return -ENXIO; 
#endif
   return bt_driver_register(bt_lower);
```

Comment (amitpurwar) [reply os/net/bluetooth/bt_netdev.c]: I agree with you and this will be designed at later stage when clarity emerges with respect to driver.

Comment (amitpurwar) [reply os/net/bluetooth/bt_netdev.c]: currently removed CONFIG_BLUETOOTH_NULL

**Comment by seinfra** [comment]
Target : [70d1d4769c49944e728f0765c47085400ea16a3d] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [70d1d4769c49944e728f0765c47085400ea16a3d] - Code Rule Check (C++) OK.
